pkg/cmd/adm/register_member.go

```diff
@@ -247,8 +261,42 @@ func newRestClient(kubeConfigPath string) (*rest.RESTClient, error) {
 	return restClient, nil
 }
 
-func generateKubeConfig(token, apiEndpoint, namespace string, insecureSkipTLSVerify bool) *clientcmdapi.Config {
-	// create apiConfig based on the secret content
+func generateKubeConfig(token, namespace string, insecureSkipTLSVerify *bool, sourceKubeConfig *clientcmdapi.Config) (*clientcmdapi.Config, error) {
+	sourceContext, present := sourceKubeConfig.Contexts[sourceKubeConfig.CurrentContext]
+	if !present {
+		return nil, fmt.Errorf("%w: current context not present", invalidKubeConfigError)
+	}
+	sourceCluster, present := sourceKubeConfig.Clusters[sourceContext.Cluster]
+	if !present {
+		return nil, fmt.Errorf("%w: cluster definition not found", invalidKubeConfigError)
+	}
+	sourceAuth, present := sourceKubeConfig.AuthInfos[sourceContext.AuthInfo]
+	if !present {
+		// can happen in tests, unlikely in practice :)
+		// The token auth will work like this though as long as there are no required client certs.
+		sourceAuth = clientcmdapi.NewAuthInfo()
+	}
+
+	// set the token in the kubeconfig and clear out any other potential auth method and impersonation
+	targetAuth := sourceAuth.DeepCopy()
+	targetAuth.Token = token
+	targetAuth.TokenFile = ""
+	targetAuth.Username = ""
+	targetAuth.Password = ""
+	targetAuth.Impersonate = ""
+	targetAuth.ImpersonateUID = ""
+	targetAuth.ImpersonateGroups = []string{}
+	targetAuth.ImpersonateUserExtra = map[string][]string{}
+	targetAuth.Exec = nil
+	targetAuth.AuthProvider = nil
```

Comment:
wasn't the original approach for `auth` section easier? I mean setting only those values that we need to set compared to deleting everything that we don't wan to have set?

Reply:
Hmm... I just wanted to use 1 approach for both auth and cluster sections (which is now also copied). I chose copying and then clearing/re-setting fields that we don't need or want to redefine.

But thinking about it further I am not 100% sure how to do this best. IMHO it is not important whether we just "pick what we want" or "clear what we don't want" but rather what fields are actually applicable. 

After thinking about it more, I am not so sure anymore that, especially in the case of cluster, the fields in the kubeconfig that the caller uses to connect to the cluster are going to be the same as the ones needed by the kubeconfig used by the SAs within the cluster. Things like `Cluster.TLSServerName` or indeed ` Cluster.CertificateAuthority` might actually be different, am I right?

I didn't want to complicate it and just use 1 approach for everything (copy and reset) but maybe we need something completely different...

Let's talk about this next year :)

Reply:
I would vote for "setting only applicable fields" which is more close to "pick what we want"
how about this one:
* set only applicable fields for auth
* copy the whole cluster context if the "insecure" flag is not define (as all CA/TLS fields are applicable in this case)
* if the "insecure" flag is used, then set only `Server` and `InsecureSkipTLSVerify`

that should be easy to read and understand. 

But TBH, either option is fine. Let's go for something that is gonna be easy to maintain & read and keep it simple, we can improve it later if needed.

Reply:
> copy the whole cluster context if the "insecure" flag is not define (as all CA/TLS fields are applicable in this case)

I don't think this is universally true. The CA/TLS can be different for communication from outside the cluster or within the cluster. We use the passed in kubeconfig to talk to the cluster from the outside but also to compose a new kubeconfig that should be used by the SA to communicate within the cluster. E.g. https://docs.openshift.com/container-platform/4.17/security/certificate_types_descriptions/user-provided-certificates-for-api-server.html vs https://docs.openshift.com/container-platform/4.17/security/certificate_types_descriptions/service-ca-certificates.html.